libraries/AP_Motors/AP_MotorsMulticopter.cpp

```diff
@@ -358,20 +313,163 @@ void AP_MotorsMulticopter::set_throttle_range(uint16_t min_throttle, int16_t rad
+                break;
+            }
 
-    // initialise maximum throttle to current throttle
-    _max_throttle = constrain_int16(_throttle_control_input, 0, AP_MOTORS_DEFAULT_MAX_THROTTLE);
+            // set and increment ramp variables
+            _throttle_low_end_pct = 0.0f;
+            _throttle_thrust_max = 0.0f;
+            _throttle_rpy_mix = 0.0f;
+            _throttle_rpy_mix_desired = 0.0f;
+            break;
+
+        case SPIN_WHEN_ARMED:{
+            // Motors should be stationary or at spin when armed.
+            // Servoes should be moving to correct the current attitude.
+
+            // set limits flags
+            limit.roll_pitch = true;
+            limit.yaw = true;
+            limit.throttle_lower = true;
+            limit.throttle_upper = true;
+
+            // set and increment ramp variables
+            float spool_step = 1.0f/(AP_MOTORS_SPOOL_UP_TIME*_loop_rate);
+            if(_spool_desired == DESIRED_SHUT_DOWN){
+                _throttle_low_end_pct -= spool_step;
+            } else if(_spool_desired == DESIRED_THROTTLE_UNLIMITED ){
+                _throttle_low_end_pct += spool_step;
+            }else{
+                _throttle_low_end_pct += constrain_float(spin_when_armed_low_end_pct()-_throttle_low_end_pct, -spool_step, spool_step);
+            }
+            _throttle_thrust_max = 0.0f;
+            _throttle_rpy_mix = 0.0f;
+            _throttle_rpy_mix_desired = 0.0f;
+
+            // constrain ramp value and update mode
+            if (_throttle_low_end_pct <= 0.0f) {
+                _throttle_low_end_pct = 0.0f;
+                _multicopter_flags.spool_mode = SHUT_DOWN;
+            } else if (_throttle_low_end_pct >= 1.0f) {
+                _throttle_low_end_pct = 1.0f;
+                _multicopter_flags.spool_mode = SPOOL_UP;
+            }
+            break;
+        }
+        case SPOOL_UP:
+            // Maximum throttle should move from minimum to maximum.
+            // Servoes should exhibit normal flight behavior.
+
+            // initialize limits flags
+            limit.roll_pitch = false;
+            limit.yaw = false;
+            limit.throttle_lower = false;
+            limit.throttle_upper = false;
+
+            // make sure the motors are spooling in the correct direction
+            if(_spool_desired != DESIRED_THROTTLE_UNLIMITED ){
+                _multicopter_flags.spool_mode = SPOOL_DOWN;
+                break;
+            }
+
+            // set and increment ramp variables
+            _throttle_low_end_pct = 1.0f;
+            _throttle_thrust_max += 1.0f/(AP_MOTORS_SPOOL_UP_TIME*_loop_rate);
+            _throttle_rpy_mix = 0.0f;
+            _throttle_rpy_mix_desired = 0.0f;
+
+            // constrain ramp value and update mode
+            if (_throttle_thrust_max >= MIN(get_throttle(), get_current_limit_max_throttle())) {
+                _throttle_thrust_max = get_current_limit_max_throttle();
```

Comment:
_throttle_thrust_max is increased until it is no longer limiting the throttle request or it is equal to the limit set by the current limiting code. At this point the copter is flying in it's normal unlimited state and the spool mode can be set to throttle unlimited.

_throttle_thrust_max  is the maximum throttle that can be used. 

So if get_throttle() < get_current_limit_max_throttle(), it will constrain the desired throttle until it is larger than get_throttle(). At this point the throttle is no longer constrained so we can set _throttle_thrust_max to get_current_limit_max_throttle() without any change or discontinuity in the output.

If get_throttle() > get_current_limit_max_throttle(), _throttle_thrust_max will reach get_current_limit_max_throttle() first. In normal operation _throttle_thrust_max = get_current_limit_max_throttle(), so while the throttle will still be constrained, we can still change to the standard unlimited mode without any change or discontinuity in the output.

Thanks for looking over this!!!!